### ISSUE TITLE
Dataset

### ISSUE THREAD

**Issue by cuongnc220592** [WAMAWAMA/TNSCUI2020-Seg-Rank1st#1]
Hi

Could you please upload the dataset of the competition
I do not have the permission to download it on the website
https://tn-scui2020.grand-challenge.org/Source_links/

Thank you very much

**Comment by WAMAWAMA** [comment]
Since the sponsor of TNSCUI challenge does not allow the dissemination of the 2020 TN-SCUI dataset, we give another thyroid ultrasound dataset demo for your reference. The demo dataset was derived from the DDTI open-access dataset, and we have processed the dataset into the format of 2020 TN-SCUI dataset，you can download it [here](https://github.com/WAMAWAMA/TNSCUI2020-Seg-Rank1st#how-to-run-this-code-on-the-tn-scui-dataset-or-your-own-custom-dataset) 😉.

**Comment by cuongnc220592** [comment]
Hi
As you mentioned that you only train with the second network, right.?
If i try to train the model from scratch with my dataset. Do I need to train both stage 1 and stage 2 models ??

**Comment by WAMAWAMA** [comment]
> Hi @WAMAWAMA
> As you mentioned that you only train with the second network, right.?
> If i try to train the model from scratch with my dataset. Do I need to train both stage 1 and stage 2 models ??

Hi @cuongnc220592 
I don't understand your question.
I trained both the 1st and 2nd networks. You can download the DDTI dataset from our link, and you will find the two folders:
 -  `DDTI\2_preprocessed_data\stage1`
 -  `DDTI\2_preprocessed_data\stage2`

the images in folder `stage1`  are for training the 1st network, and the images in folder `stage2` are for training the 2nd one.

And, if you want to train the model from scratch with your dataset, **of course** you need to train both stage 1 and stage 2 models, respectively. The only and necessary difference between training the two networks is the training dataset. The full image is used to train the first network for locating, 'ROI' image is used to train the second network for refining.

**Comment by cuongnc220592** [comment]
Thanks for your help @WAMAWAMA 
This sentence make me confused. But I understand now.
> Our preliminary experiments show that the provided context information in first network may do not play a significant auxiliary role for refinement of the second network. Therefore, **we only train the second network** using images within region of interest (ROI) obtained from ground truth (GT).

 Could you explain to me about the training network?  Did you use the original UNET architecture  for both stage1 and stage2 ?
I do not see you mention it on the README.
Thanks you very much

**Comment by WAMAWAMA** [comment]
@cuongnc220592 
😮 I am sorry that I did not describe the full training details in the readme.
We use the `DeeplabV3+ with efficientnet-B6 encoder`model for both stage1 and stage2, which implemented by [`smp`](https://github.com/qubvel/segmentation_models.pytorch) Lib.  
And [`smp`](https://github.com/qubvel/segmentation_models.pytorch) actually contains Unet structure as well, but it's different from the original structure, because [`smp`](https://github.com/qubvel/segmentation_models.pytorch) takes some network structures with pre-trained weights as encoder, and this leads to some performance improvements relative to the original Unet structure. (BTW, we also tried original Unet or Unet++ , based on this repo: [https://github.com/4uiiurz1/pytorch-nested-unet](https://github.com/4uiiurz1/pytorch-nested-unet)

**Comment by cuongnc220592** [comment]
I have a question for training second network.
Do I need set image_size to 512 for second network? You do not mentioncin README.
Thanks you

**Comment by WAMAWAMA** [comment]
@cuongnc220592 
sorry for confusing you， i'll modify the `README` soon😉
(BTW, both 512 and 256 are OK, There is little difference in performance between the two image_size settings)
and for the **'false positive’** problem，the process about "get the largest connected components" was designed for THIS CHALLENGE, so it is only suitable for 'Single target sementation'. 
In my experience, there may be 2 ways:
 1.multi-task learning with segmentation and classification(e.g. image has nodule or not);
 2. segmenting the background but not nodules, may add a post classification network to classify the predicted mask

**Comment by cuongnc220592** [comment]
Dear @WAMAWAMA 

Dear author,
I have a weird problem in training.
I use private dataset, and modify your code  like this:

   `train_img_folder = '../../2_TNSCUI2020-Seg-Rank1st/preprocessed_data/train/stage1/p_image'
    train_mask_folder = '../../2_TNSCUI2020-Seg-Rank1st/preprocessed_data/train/stage1/p_mask'
    test_img_folder = '../../2_TNSCUI2020-Seg-Rank1st/preprocessed_data/val/stage1/p_image'
    test_mask_folder = '../../2_TNSCUI2020-Seg-Rank1st/preprocessed_data/val/stage1/p_mask' 

train_list = get_filelist_frompath(train_img_folder,'jpg') 
    train_list_GT = [train_mask_folder+sep+i.split(sep)[-1] for i in train_list]
    test_list = get_filelist_frompath(test_img_folder,'jpg') 
    test_list_GT = [test_mask_folder+sep+i.split(sep)[-1] for i in test_list]
`

In the first training time of first model, everything is OK with the loss  decrease and IOU, DICE score increase each epoch.
Epoch [1/405], Loss: 0.8821
[Validation] Acc: 0.4715, SE: 0.5164, SP: 0.4716, PC: 0.0447, Dice: 0.0723, IOU: 0.0395
Best model in epoch 1, score : 0.0395
Epoch [2/405], Loss: 0.6939
[Validation] Acc: 0.8723, SE: 0.9429, SP: 0.8736, PC: 0.3950, Dice: 0.4846, IOU: 0.3708
Best model in epoch 2, score : 0.3708
Epoch [3/405], Loss: 0.4919
[Validation] Acc: 0.9625, SE: 0.9125, SP: 0.9652, PC: 0.5708, Dice: 0.6609, IOU: 0.5300
Best model in epoch 3, score : 0.5300

 However, when I try to retrain model in second time, everything (data model, and configuration) is similar. the accuracy at the first step is very high at the first epoch, but the IOU and DICE do not improve.
Epoch [1/405], Loss: 0.8837
[Validation] Acc: 0.8234, SE: 0.1811, SP: 0.9955, PC: 0.1811, Dice: 1.9727, IOU: 0.1811
Best model in epoch 1, score : 0.1811
Epoch [2/405], Loss: 0.6891
[Validation] Acc: 0.9425, SE: 0.0991, SP: 0.9585, PC: 0.0991, Dice: 1.4994, IOU: 0.0991.
Epoch [3/405], Loss: 0.4433
[Validation] Acc: 0.9780, SE: 0.0623, SP: 0.9596, PC: 0.0623, Dice: 1.2847, IOU: 0.0623

Iam using torchvision= 0.8.1, pytorch= 1.2.0

Could you give me some suggestion to solve the problem.
Thank you very much

**Comment by WAMAWAMA** [comment]
@cuongnc220592 
Hi cuongnc220592 , sorry for replying so late 😂
The cause of this problem may be the torch version. 
Because DICE and IoU was calculated with torch tensor when training, torch version may make this process wrong, this bug was also found in this  issue #8  
You can modify the code of calculating DICE and IoU with numpy to remove this bug👍
ill try to remove this bug in few days

**Comment by cuongnc220592** [comment]
Hi @WAMAWAMA 
I have one more concern that I have a dataset with image size 512x512 and small segmented objects (~30 mm).
Hence, I would like train stage_1 model on the same image size (512x512) to save the information of small objects.
In step_2 model I crop the object to 512x512 with original model.
Do my idea can implement in your code?
Do I need to change anything to implement (image_size at stage_1, etc ...)
Thank you so much.
Best.

**Comment by WAMAWAMA** [comment]
Actually, your idea has already be implemented in this code, in the [step3_TestOrInference.py](https://github.com/WAMAWAMA/TNSCUI2020-Seg-Rank1st/blob/master/step2to4_train_validate_inference/step3_TestOrInference.py)
you just need to:
1) train the two different stage model, stage 1 with the full-size images in the size of 512 x 512 pixels, stage 2 with cropped and resized images of "small segmented objects" in the size of 512 x 512 pixels, 
2) test by changing the params  `c1_size ` and `c1_size ` to 512 in step3_TestOrInference.py.

**Comment by cuongnc220592** [comment]
How about the result when you implement with same image size in both models?
Why you decide to reduce the image size at stage 1 ?

**Comment by WAMAWAMA** [comment]
1) DSC will go down a little bit when using the same input size, but I think that's because of dataset specificity, you may get different results in other datasets.
2) I decide to reduce the image size at stage 1 because stage 1 is just trained for localization, so there's no need to feed a huge and high-resolution image to train it.
